nft-simple/src/nft_core.rs

```diff
@@ -60,54 +78,121 @@ impl NonFungibleTokenCore for Contract {
         &mut self,
         receiver_id: ValidAccountId,
         token_id: TokenId,
+        enforce_owner_id: Option<ValidAccountId>,
         memo: Option<String>,
     ) {
         assert_one_yocto();
         let sender_id = env::predecessor_account_id();
-        self.internal_transfer(&sender_id, receiver_id.as_ref(), &token_id, memo);
+        let (previous_owner_id, approved_account_ids) = self.internal_transfer(
+            &sender_id,
+            receiver_id.as_ref(),
+            &token_id,
+            enforce_owner_id.as_ref(),
+            memo,
+        );
+        refund_approved_account_ids(previous_owner_id, &approved_account_ids);
     }
 
     #[payable]
     fn nft_transfer_call(
         &mut self,
         receiver_id: ValidAccountId,
         token_id: TokenId,
-        msg: String,
+        enforce_owner_id: Option<ValidAccountId>,
         memo: Option<String>,
+        msg: String,
     ) -> Promise {
         assert_one_yocto();
         let sender_id = env::predecessor_account_id();
-        self.internal_transfer(&sender_id, receiver_id.as_ref(), &token_id, memo);
+        let (owner_id, approved_account_ids) = self.internal_transfer(
+            &sender_id,
+            receiver_id.as_ref(),
+            &token_id,
+            enforce_owner_id.as_ref(),
+            memo,
+        );
         // Initiating receiver's call and the callback
         ext_non_fungible_token_receiver::nft_on_transfer(
             sender_id.clone(),
+            owner_id.clone(),
             token_id.clone(),
             msg,
             receiver_id.as_ref(),
             NO_DEPOSIT,
             env::prepaid_gas() - GAS_FOR_NFT_TRANSFER_CALL,
         )
         .then(ext_self::nft_resolve_transfer(
-            sender_id,
+            owner_id,
             receiver_id.into(),
+            approved_account_ids,
             token_id,
             &env::current_account_id(),
             NO_DEPOSIT,
             GAS_FOR_RESOLVE_TRANSFER,
         ))
     }
 
+    #[payable]
+    fn nft_approve_account_id(&mut self, token_id: TokenId, account_id: ValidAccountId) -> bool {
+        let mut token = self.tokens_by_id.get(&token_id).expect("Token not found");
```

Comment:
I think this would be a likely attack vector and we'd want:

    assert_one_yocto();

Reply:
Good catch